docs/common/use-table-of-contents.ts

```diff
@@ -0,0 +1,109 @@
+export function useTableOfContents(options: TableOfContentsOptions = {}) {
+  const [activeId, setActiveId] = useState<string>();
+  const headings = useHeadings(options.root, options.selector);
+  const router = useRouter();
+
+  useObserver(options.root, setActiveId, headings);
+
+  const onRouteChange = useCallback(
+    (url: string) => setActiveId(url.split('#', 2).pop()!),
+    [setActiveId]
+  );
+
+  useEffect(function onDidMount() {
+    router.events.on('hashChangeStart', onRouteChange);
+    if (window.location.hash) {
+      onRouteChange(window.location.hash);
+    }
+    return () => router.events.off('hashChangeStart', onRouteChange);
+  }, []);
+
+  return { headings, activeId };
+}
+
+/**
+ * Find the heading elements within a root element.
+ */
+function useHeadings(root: Element | undefined, selector = 'h2, h3') {
+  const [headings, setHeadings] = useState<HTMLHeadingElement[]>([]);
+
+  useEffect(
+    function onDidMount() {
+      const elements = root?.querySelectorAll<HTMLHeadingElement>(selector);
+      if (elements?.length) {
+        setHeadings(Array.from(elements));
+      } else {
+        setHeadings([]);
+      }
+    },
+    [root, selector]
+  );
+
+  return headings;
+}
+
+/**
+ * Observe elements to detect them being scrolled in and out of view.
+ * This uses an intersection observer under the hood.
+ */
+function useObserver(
+  root: Element | undefined,
+  onVisible: (id: string) => any,
+  observerables: Element[]
+) {
+  // TODO(cedric): find a better way to detect changes of the observerables
+  const observerablesHash = observerables.map(el => findHeadingId(el)).join('');
+
+  const onObserve = useCallback(
+    (entries: IntersectionObserverEntry[]) => {
+      const visible = entries.filter(entry => entry.isIntersecting);
+      if (visible.length === 1) {
+        // TODO(cedric): move the heading id back the the heading element
+        onVisible(findHeadingId(visible[0].target));
+      } else if (visible.length > 1) {
+        const sorted = visible.sort(
+          (a, b) => observerables.indexOf(a.target) - observerables.indexOf(b.target)
+        );
+        onVisible(findHeadingId(sorted[0].target));
+      }
+    },
+    [onVisible, observerablesHash]
+  );
+
+  useEffect(
+    function onDidMount() {
+      const observer = new IntersectionObserver(onObserve, {
+        root,
+        // TODO(cedric): make sure these margins are tweaked properly with the new heading components
+        rootMargin: '-25% 0px -25% 0px',
+      });
+      observerables.forEach(observerable => observer.observe(observerable));
+      return () => observer.disconnect();
+    },
+    [root, onVisible, onObserve, observerablesHash]
+  );
+}
+
+/**
+ * Find the heading ID of a heading.
+ * @todo move the heading id back the the heading element, making this function obsolete
+ */
+export function findHeadingId(element: Element) {
+  return Array.from(element.getElementsByTagName('span')).find(span => span.id)!.id;
```

Comment:
I blame the `Permalink` for this. When we have new heading components, this should be obsolete, should be able to use `heading[id]` instead. And when we do need something like that, add maybe `heading[data-hash]` or whatever to get rid of unnecessary iterations.